cranelift/bitset/src/scalar.rs

```diff
@@ -0,0 +1,575 @@
+    ///
+    /// // The bounds must fit within the backing scalar type.
+    /// let bitset = ScalarBitSet::<u64>::from_range(3, 69);
+    /// ```
+    #[inline]
+    pub fn from_range(lo: u8, hi: u8) -> Self {
+        assert!(lo <= hi);
+        assert!(hi <= Self::capacity());
+
+        let one = T::from(1);
+
+        // We can't just do (one << hi) - one here as the shift may overflow
+        let hi_rng = if hi >= 1 {
+            (one << (hi - 1)) + ((one << (hi - 1)) - one)
+        } else {
+            T::from(0)
+        };
+
+        let lo_rng = (one << lo) - one;
+
+        Self(hi_rng - lo_rng)
+    }
+
+    /// The maximum number of bits that can be stored in this bitset.
+    ///
+    /// If you need more bits than this, use a
+    /// [`CompoundBitSet`][crate::CompoundBitSet] instead of a `ScalarBitSet`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use cranelift_bitset::ScalarBitSet;
+    ///
+    /// assert_eq!(ScalarBitSet::<u8>::capacity(), 8);
+    /// assert_eq!(ScalarBitSet::<u64>::capacity(), 64);
+    /// ```
+    #[inline]
+    pub fn capacity() -> u8 {
+        u8::try_from(size_of::<T>()).unwrap() * 8
+    }
+
+    /// Get the number of elements in this set.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use cranelift_bitset::ScalarBitSet;
+    ///
+    /// let mut bitset = ScalarBitSet::<u64>::new();
+    ///
+    /// assert_eq!(bitset.len(), 0);
+    ///
+    /// bitset.insert(24);
+    /// bitset.insert(13);
+    /// bitset.insert(36);
+    ///
+    /// assert_eq!(bitset.len(), 3);
+    /// ```
+    #[inline]
+    pub fn len(&self) -> u8 {
+        self.0.count_ones()
+    }
+
+    /// Is this bitset empty?
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use cranelift_bitset::ScalarBitSet;
+    ///
+    /// let mut bitset = ScalarBitSet::<u16>::new();
+    ///
+    /// assert!(bitset.is_empty());
+    ///
+    /// bitset.insert(10);
+    ///
+    /// assert!(!bitset.is_empty());
+    /// ```
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.0 == T::from(0)
+    }
+
+    /// Check whether this bitset contains `i`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use cranelift_bitset::ScalarBitSet;
+    ///
+    /// let mut bitset = ScalarBitSet::<u8>::new();
+    ///
+    /// assert!(!bitset.contains(7));
+    ///
+    /// bitset.insert(7);
+    ///
+    /// assert!(bitset.contains(7));
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// Panics if `i` is greater than or equal to [`Self::capacity()`][Self::capacity].
+    ///
+    /// ```should_panic
+    /// use cranelift_bitset::ScalarBitSet;
+    ///
+    /// let mut bitset = ScalarBitSet::<u8>::new();
+    ///
+    /// // A `ScalarBitSet<u8>` can only hold the elements `0..=7`, so `8` is
+    /// // out of bounds and will trigger a panic.
+    /// bitset.contains(8);
+    /// ```
+    #[inline]
+    pub fn contains(&self, i: u8) -> bool {
+        assert!(i < Self::capacity());
```

Comment:
Given the perf-sensitive nature of bitsets should this be `debug_assert!`?

Reply:
I was thinking that originally, but figured we might as well do the more-correct thing by default and add `unchecked` versions later if profiling shows that these asserts actually matter.

I kind of expect LLVM to clean them all up, at least when they are called via `CompoundBitSet` methods, since it should be able to determine the ranges of values flowing into these methods based on the modulo operation in `CompoundBitSet::word_and_bit`.